rfc/monorepo.md

```diff
@@ -0,0 +1,557 @@
+1. With heavy builds comes longer CI times.  Some intelligence will need to be added to the Github Actions workflows to select Gradle tasks intelligently based on the incoming diff.
+    - We use this kind of functionality heavily in our own forks/extensions - in combination with the cache, our CI times are ~30 minutes for a full build and test of all Spinnaker code.  Code that does not touch `kork` can omit many tests, and takes between 5-15 minutes.
+1. Libraries still need to be published, even though the project itself no longer uses them - there are many external consumers.
+
+### Monorepo Github Location
+
+The monorepo will live at `github.com/spinnaker/spinnaker`.  All other repositories will be archived once the transition is completed.
+
+### Primary/Default Branch Name
+
+The primary/default branch of the monorepo will be renamed from `master` to `main`.
+
+### Yes, Deck Too
+
+This design explicitly integrates `deck` and `deck-kayenta` into the monorepo.  Cross-cutting changes do occur between the UI and the backend, and Deck is already a [Lerna](https://lerna.js.org) monorepo within itself - building all of it from its existing Gradle entry point is simple.
+
+Linking the builds of `deck`, and `deck-kayenta` is not something Gradle can do, but we have two good options here:
+
+1. Use the local `deck-kayenta` code during the build process of `deck`, using `yarn`'s workspaces feature
+1. Just integrate `deck-kayenta` as a regular `deck` module, like all the other Lerna subprojects under `packages/*`
+
+### The `spin` CLI
+
+This project is actually the most difficult to integrate into the monorepo, as golang builds tend to be really finicky about their environment.  I think Github Actions should afford us the flexibility to partition that out reasonably, but that requires some investigative work.
+
+There are definitely concrete benefits to including it in the build cycle - it can directly benefit from using adjacent Gate libraries, and it's much more testable when colocated against the APIs it is calling.  Setting up ephemeral CLI testing environments (on every change to Gate, even), is well within GHA's capabilities.
+
+### Halyard
+
+The inclusion of Halyard into the monorepo merits some discussion.  It is a tool everyone wants to see replaced, but has no replacement yet.  It consumes many Spinnaker service libraries, and would greatly benefit from composite builds.
+
+However, it sits in the unique position of not really needing to be up to date with its Spinnaker dependencies.  We need to weigh the benefits of building Halyard more conveniently against failing pull request builds because a contributor broke something in Halyard.  We should also consider that Halyard being stale gets harder and harder to fix over time, so perhaps the possibility of disruption to pull request builds is worth it in the long run.
```

Comment:
I'd like to include halyard in the monorepo.  Catching breakage during e.g. spring boot upgrades (and gradle upgrades, java, etc.) would be super helpful.  It's true that it doesn't need to be up to date, but it may as well be.